src/python/pants/backend/docker/util_rules/docker_build_context.py

```diff
@@ -66,6 +71,33 @@ def merge(self, other: Mapping[str, Mapping[str, str]]) -> DockerVersionContext:
         return DockerVersionContext.from_dict({**self, **other})
 
 
+class DockerContextFilesAcceptableInputsField(ABC, SourcesField):
+    """This is a meta field for the context files generator, to tell the codegen machinery what
+    source fields are good to use."""
+
+
+DockerContextFilesAcceptableInputsField.register(FileSourceField)
+DockerContextFilesAcceptableInputsField.register(ShellSourceField)
+
+
+class DockerContextFilesSourcesField(SourcesField):
+    """This is just a type marker for the codegen machinery."""
+
+
+class GenerateDockerContextFiles(GenerateSourcesRequest):
+    """This translates all files from acceptable Source fields for the docker context using the
+    `codegen` machinery."""
+
+    input = DockerContextFilesAcceptableInputsField
+    output = DockerContextFilesSourcesField
+
+
+@rule
+async def hydrate_input_sources(request: GenerateDockerContextFiles) -> GeneratedSources:
+    # We simply pass the files on, as-is
+    return GeneratedSources(request.protocol_sources)
```

Comment:
With this technique you can actually write any number of rules for any set of input fields, as long as you have dedicated output types.